File: baekjoon/gold/11054/11054.py
n = int(input())

arr = list(map(int, input().split()))
dp = [0] * n
for i in range(n-1):
    for j in range(i+1, n):
        if arr[i] < arr[j]:
            dp[j] = max(dp[i]+1, dp[j])
dp1 = [0] * n
for i in range(n-1):
    for j in range(i+1, n):
        if arr[-i-1] < arr[-j-1]:
            dp1[-j-1] = max(dp1[-i-1]+1, dp1[-j-1])
print(dp, dp1)
result = [dp[i]+dp1[i] for i in range(n)]
print(max(result)+1)

File: baekjoon/gold/11054/11054.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	reader *bufio.Reader = bufio.NewReader(os.Stdin)
	writer *bufio.Writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var n int

	fmt.Fscanf(reader, "%d\n", &n)
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d ", &arr[i])
	}

	dp1 := make([]int, n)
	dp2 := make([]int, n)
	for i := 0; i < n; i++ {
		dp1[i] = 1
		for j := 0; j < i; j++ {
			if arr[j] < arr[i] && dp1[i] < dp1[j]+1 {
				dp1[i] = dp1[j] + 1
			}
		}
	}

	for i := n - 1; i >= 0; i-- {
		dp2[i] = 1
		for j := n - 1; j > i; j-- {
			if arr[j] < arr[i] && dp2[i] < dp2[j]+1 {
				dp2[i] = dp2[j] + 1
			}
		}
	}

	max := 0
	for i := 0; i < n; i++ {
		if max < dp1[i]+dp2[i]-1 {
			max = dp1[i] + dp2[i] - 1
		}
	}

	fmt.Fprintln(writer, max)
}
